baiji/serialization/json.py

```diff
@@ -23,83 +24,140 @@ def _load(f, *args, **kwargs):
 def loads(*args, **kwargs):
     return json.loads(*args, **_load_args(kwargs))
 
-def _load_args(kwargs):
-    kwargs.update(object_hook=_json_decode)
-    return kwargs
+class MethodListCaller(object):
+    def register(self, method, index=-1):
+        if not hasattr(self, 'method_list'):
+            self.clear()
+        if index == -1:
+            index = len(self.method_list)
+        self.method_list.insert(index, method)
 
-def _dump_args(kwargs):
-    if kwargs.get('primitive', False):
-        kwargs.update(default=_json_encode_primitive)
+    def clear(self):
+        # be defensive if someone forgets to call super pylint: disable=attribute-defined-outside-init
+        self.method_list = []
+
+    def __call__(self, x):
```

Comment:
That's basically what's going on here. The acutal computation is being delegated to the registered methods (the various `encode_*` and `decode_*` in the subclasses). I'm a fan of moving most of the code over like this, but not of making `__call__` simply be an alias of another function, because then it ceases to be clear to the user how they should be calling it. The only time I agree with that is when `__call__` is being added as syntactic sugar. In this case `__call__` is the primary interface, forced on us by the fact that simplejson is expected a function, not an object and we're faking it out.